pkgs/checks/lib/src/checks.dart

```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
   Context<T> get context => _context;
 }
 
-/// The expectation and nesting context already applied to a [Subject].
+/// The context for a [Subject] that allows asserting expectations and creating
+/// nested subjects.
 ///
-/// This is the surface of interaction for expectation extension method
-/// implementations.
+/// Expectation extension methods use [ContextExtension] to get access to the
+/// subject's [Context]. Expectations pass callbacks which describe what is
+/// checked, and perform the expectation, respectively.
 ///
-/// The `expect` and `expectAsync` can test the value and optionally reject it.
-/// The `nest` and `nestAsync` can test the value, and also extract some other
-/// property from it for further checking.
+/// When all expectations on a subject succeed, the descriptions will not be
+/// needed and the callbacks will not be called. Similarly, if an expectation
+/// fails in a [softCheck] or [softCheckAsync], the descriptions may not be
+/// used and the callbacks will not be called. Where possible, string formatting
+/// for the descriptions should be performed in the callback, not ahead of time.
+///
+/// When an expectation is used on a [ConditionSubject] and passed to
+/// [describe], there is no value to check an expectaion against, so only the
+/// description callbacks are used.
+///
+/// When both callbacks are used, the description callback will always be called
+/// strictly after the expectation callback is called.
```

Comment:
> Do you have, or expect, code which depends on that order?

Yes. Sometimes where I can't do `async` work for the description callback, I might have a fallback description by default, but gather some async information during the predicate and format it for a richer description when possible.

For example the `inOrder` extension method cannot synchronously describe the `Condition` instances it will check because they will use async expectations. It has a fallback for `describe` calling out the number of conditions that need to be satisfied, but only shows their details when the actual has run.

https://github.com/dart-lang/test/blob/911659bbc3eea9c133263b2cfcaacdc376767faf/pkgs/checks/lib/src/extensions/async.dart#L221-L224

> Such code would likely maintain internal state which makes it impossible to replay the same check/description on a separate value, which would make me very disappointed.

That is the case. We should investigate whether there is a different way to provide a good UX for this.
Maybe we can devise an explicit way for the expectation callback to communicate to the description callback that would be compatible with reuse. Or maybe we should sacrifice the UX for those use cases.